src/Language/Fixpoint/Smt/Interface.hs

```diff
@@ -255,41 +267,47 @@ negativeP
        return $ "(" <> v <> ")"
 
 -- | Writes a line of input for the SMT solver and to the log if there is one.
-smtWriteRaw :: Context -> Raw -> IO ()
-smtWriteRaw me !s = {- SCC "smtWriteRaw" -} do
+smtWriteRaw :: Context -> Raw -> Bool -> IO ()
+smtWriteRaw me !s expectResponse = {- SCC "smtWriteRaw" -} do
   -- whenLoud $ do LTIO.appendFile debugFile (s <> "\n")
   --               LTIO.putStrLn ("CMD-RAW:" <> s <> ":CMD-RAW:DONE")
-  hPutStrLnNow (ctxIn me) s
   maybe (return ()) (`LTIO.hPutStrLn` s) (ctxLog me)
+  -- TODO don't rely on Text
+  let sendWith sender = sender (ctxSolver me) $ lazyByteString $ LTE.encodeUtf8 s
+  if expectResponse
+    then do
+      resp <-
+        LBS.reverse . LBS.dropWhile isSpace . LBS.reverse
+          <$> sendWith Bck.command
+      modifyIORef (ctxResp me) (<> (resp <> "\n"))
+    else do
+      _ <- sendWith Bck.command_
+      return ()
 
 -- | Reads a line of output from the SMT solver.
 smtReadRaw :: Context -> IO T.Text
 smtReadRaw me = do
-  eof <- hIsEOF (ctxOut me)
-  if eof then Misc.errorstar "SMT returned End of File" else
-    TIO.hGetLine (ctxOut me)
-{-# SCC smtReadRaw  #-}
-
-hPutStrLnNow :: Handle -> LT.Text -> IO ()
-hPutStrLnNow h !s = LTIO.hPutStrLn h s >> hFlush h
-{-# SCC hPutStrLnNow #-}
+  respLn <- atomicModifyIORef (ctxResp me) $ \resps ->
```

Comment:
What is the need to call `atomicModifyIORef` here?

Reply:
None, I probably wrote this at a time when I wasn't sure of the concurrency model of liquid-fixpoint yet.

Reply:
Removed in https://github.com/ucsd-progsys/liquid-fixpoint/pull/641/commits/cb9ff514fbcbe3f1c8ec08997391620d0871feb5.